[호돌] API 테스트/문서자동화 미션 제출합니다.

woonjangahn opened the pull request:
잘 부탁드립니다 🚀

woonjangahn committed:
feat : 문서 자동화 실습 진행

- Spring REST Docs 의존성 추가
- MemberControllerTest 내용 복사/붙여넣기
- AuthAcceptanceTest 임시 비활성화
- LineControllerTest를 SpringBootTest로 변경
- MemberDocumentation 주석 해제

woonjangahn committed:
feat : 사용자 인증 실습 진행

- Utf8EncodingFilter 작성
- Session / Basic / Bearer AuthInterceptor 작성
- MockMvc 사용하여 AuthAcceptanceTest 작성

woonjangahn committed:
docs : 요구사항 작성

woonjangahn committed:
feat : 회원 정보 관리 인수테스트 작성

- JWT 토큰 활용하여 회원정보 CRUD시 인증절차 거치도록 구현
- 인증이 필요한 컨트롤러들의 패턴을 BearerAuthInterceptor에 추가
- 인증정보(LoginMember)가 없을 경우 200 상태로 빈 멤버를 응답
- 인증정보가 틀릴 경우 401 응답
- 사용하지 않는 다른 인증 및 관련 테스트 삭제
- 서버 프로젝트 스코프 컨벤션 리팩토링 진행

woonjangahn committed:
feat : 회원 CRUD 단위 테스트 작성

woonjangahn committed:
refactor : RestAssured 테스트 픽스쳐들을 MockMvc 픽스처들로 대체

- RestAssured 의존성 삭제

woonjangahn committed:
feat : 회원관리 컨트롤러 테스트 작성 및 문서 자동화

woonjangahn committed:
refactor : MockMvc 테스트 픽스쳐 리팩토링

woonjangahn committed:
etc : 프론트엔드 샘플 코드 추가 및 1단계 요구사항 반영

woonjangahn committed:
feat : Favorite CRD 인수, 단위 테스트 및 구현

woonjangahn committed:
feat : Favorite 컨트롤러 테스트 작성 및 문서화

woonjangahn committed:
feat : Favorite 페이지 연동

- Login.js에서 비밀번호 input keypress event에 onLogin 추가

woonjangahn committed:
refactor : 컨벤션에 맞게 프로젝트 스코프 리팩토링

woonjangahn committed:
refactor : 컨트롤러에서의 사용자 검증을 LoginMemberMethodArgumentResolver, LoginMember.Type으로 이동

woonjangahn committed:
refactor : GlobalExceptionHandler 및 커스텀 예외처리 추가

woonjangahn committed:
test : 일부 AcceptanceTest 다이나믹 테스트 적용 및 Fixture 리팩토링

woonjangahn committed:
docs : 요구사항 업데이트

young891221 commented on src/main/java/wooteco/subway/web/exceptions/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,27 @@
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+@RestControllerAdvice
+public class GlobalExceptionHandler {
```
디폴트 예외처리도 추가되었으면 좋겠네요ㅎ

woonjangahn replied:
```java
@ExceptionHandler(Exception.class)
public ResponseEntity<ErrorResponse> handleExceptions(InvalidAuthenticationException exception) {
    ErrorResponse errorResponse = new ErrorResponse(HttpStatus.INTERNAL_SERVER_ERROR.value(), exception);
    return ResponseEntity.status(errorResponse.getStatusCode()).body(errorResponse);
}
```
커스텀 정의된 예외 이외의 모든 디폴트 예외를 처리하도록 핸들러 메소드를 추가했습니다!

young891221 commented on build.gradle:
```diff
@@ -1,31 +1,50 @@
-	runtimeOnly 'com.h2database:h2'
+    implementation 'org.springframework.boot:spring-boot-starter-web'
+    implementation 'org.springframework.boot:spring-boot-starter-data-jdbc'
+    implementation 'net.rakugakibox.spring.boot:logback-access-spring-boot-starter:2.7.1'
+    implementation 'pl.allegro.tech.boot:handlebars-spring-boot-starter:0.3.0'
+    implementation 'org.jgrapht:jgrapht-core:1.0.1'
+    implementation 'io.jsonwebtoken:jjwt:0.9.1'
+    asciidoctor 'org.springframework.restdocs:spring-restdocs-asciidoctor:2.0.4.RELEASE'
+    testImplementation 'org.springframework.restdocs:spring-restdocs-mockmvc:2.0.4.RELEASE'
```
의존성 버전관리에 대한 부분을 생각해 주세요~!
`io.spring.dependency-management`와 관련지어 알아보면 좋을듯 합니다ㅎ

woonjangahn replied:
```groovy
asciidoctor 'org.springframework.restdocs:spring-restdocs-asciidoctor'
testImplementation 'org.springframework.restdocs:spring-restdocs-mockmvc'
```

`spring-restdocs-asciidoctor`과  `spring-restdocs-mockmvc`의 버전이 `io.spring.dependency-management`에서 관리되도록 변경했습니다!

young891221 commented on src/test/java/wooteco/subway/web/member/MemberControllerTest.java:
```diff
@@ -1,5 +1,141 @@
+
+@ExtendWith(RestDocumentationExtension.class)
+@WebMvcTest(controllers = MemberController.class)
+@Import({AuthorizationExtractor.class, JwtTokenProvider.class})
```
인증 실패, 잘못된 요청등 예외에 대한 테스트도 추가해보면 어떨까요?

woonjangahn replied:
회원가입, 로그인등에서의 인증실패와 잘못된 요청에 대한 컨트롤러 테스트를 추가하고 문서화까지 완료했습니다!

young891221 reviewed (changes requested):
안녕하세요, 호돌!
미션 구현하느라 고생 많으셨습니다ㅎㅎ
구현부터 테스트까지 깔끔해서 간단한 코멘트만 반영하고 머지해도 좋아보입니다. 확인 부탁드립니다~!

woonjangahn committed:
refactor : 1차 리뷰 반영

- GlobalExceptionHandler 디폴트 예외처리 추가
- spring.dependency-management에서 restdocs의 버전을 관리하도록 변경
- 회원가입, 로그인 등의 인증실패와 잘못된 요청에 대한 컨트롤러 테스트 추가 및 문서화

young891221 commented on src/main/java/wooteco/subway/domain/member/Member.java:
```diff
@@ -20,18 +22,23 @@ public Member() {
     }
 
+    private void validate(final String email, final String name, final String password) {
+        if (StringUtils.isBlank(email) || StringUtils.isBlank(name) || StringUtils.isBlank(password)) {
```
실제로는 각각 어떤 값이 빈 값인지 메시지를 정확히 찍어주는게 디버깅 혹은 api 요청에 도움이 될 것 같네요ㅎ

woonjangahn replied:
반영하도록 하겠습니다 :)
감사합니다!

young891221 reviewed (approved):
제가 원하는 반향으로 깔끔하게 반영 해 주셨네요. 💯 
부족한 부분이 많이 없으셔서 리뷰를 많이 못드려 아쉽습니다ㅎㅎ
이번 단계는 여기서 머지하겠습니다~!
